pkg/products/grafana/reconciler.go

```diff
@@ -0,0 +1,320 @@
+	return nil
+}
+
+func (r *Reconciler) VerifyVersion(installation *integreatlyv1alpha1.RHMI) bool {
+	return version.VerifyProductAndOperatorVersion(
+		installation.Status.Stages[integreatlyv1alpha1.ProductsStage].Products[integreatlyv1alpha1.ProductGrafana],
+		string(integreatlyv1alpha1.VersionGrafana),
+		string(integreatlyv1alpha1.OperatorVersionGrafana),
+	)
+}
+
+func NewReconciler(configManager config.ConfigReadWriter, installation *integreatlyv1alpha1.RHMI, mpm marketplace.MarketplaceInterface, recorder record.EventRecorder) (*Reconciler, error) {
+	ns := installation.Spec.NamespacePrefix + defaultInstallationNamespace
+	config, err := configManager.ReadGrafana()
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve threescale config: %w", err)
+	}
+	if config.GetNamespace() == "" {
+		config.SetNamespace(ns)
+		configManager.WriteConfig(config)
+	}
+	if config.GetOperatorNamespace() == "" {
+		if installation.Spec.OperatorsInProductNamespace {
+			config.SetOperatorNamespace(config.GetNamespace())
+		} else {
+			config.SetOperatorNamespace(config.GetNamespace() + "-operator")
+		}
+	}
+
+	logger := logrus.NewEntry(logrus.StandardLogger())
+	return &Reconciler{
+		ConfigManager: configManager,
+		Config:        config,
+		installation:  installation,
+		mpm:           mpm,
+		logger:        logger,
+		Reconciler:    resources.NewReconciler(mpm),
+		recorder:      recorder,
+	}, nil
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, installation *integreatlyv1alpha1.RHMI, product *integreatlyv1alpha1.RHMIProductStatus, client k8sclient.Client) (integreatlyv1alpha1.StatusPhase, error) {
+	logrus.Infof("Start Grafana reconcile")
+
+	operatorNamespace := r.Config.GetOperatorNamespace()
+	productNamespace := r.Config.GetNamespace()
+
+	phase, err := r.ReconcileFinalizer(ctx, client, installation, string(r.Config.GetProductName()), func() (integreatlyv1alpha1.StatusPhase, error) {
+		phase, err := resources.RemoveNamespace(ctx, installation, client, productNamespace)
+		if err != nil || phase != integreatlyv1alpha1.PhaseCompleted {
+			return phase, err
+		}
+
+		phase, err = resources.RemoveNamespace(ctx, installation, client, operatorNamespace)
+		if err != nil || phase != integreatlyv1alpha1.PhaseCompleted {
+			return phase, err
+		}
+
+		return integreatlyv1alpha1.PhaseCompleted, nil
+	})
+	if err != nil || phase != integreatlyv1alpha1.PhaseCompleted {
+		events.HandleError(r.recorder, installation, phase, "Failed to reconcile finalizer", err)
+		return phase, err
+	}
+
+	phase, err = r.ReconcileNamespace(ctx, operatorNamespace, installation, client)
+	if err != nil || phase != integreatlyv1alpha1.PhaseCompleted {
+		events.HandleError(r.recorder, installation, phase, fmt.Sprintf("Failed to reconcile %s ns", operatorNamespace), err)
+		return phase, err
+	}
+
+	err = r.createSecrets(ctx, client, installation, &grafanav1alpha1.Grafana{})
+	if err != nil || phase != integreatlyv1alpha1.PhaseCompleted {
```

Comment:
I think we should either remove the phase check here and return a new phase or have the `createSecrets` function return a phase failure rather than a standard err. if it fails to create the secret it's going to rerun the previous phase from above as it's not updated here.

Reply:
I have updated this to reconcile and report phase failed.